src/librustc_mir/const_eval.rs

```diff
@@ -370,6 +371,16 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
         }))
     }
 
+    fn call_extra_fn(
+        _ecx: &mut InterpCx<'mir, 'tcx, Self>,
+        fn_val: !,
```

Comment:
So, my prior understanding was that the never (`!`) type is never realized (and thus is the "return type" of functions that don't return, like `panic!`), so I'm confused about how we can use it as a function argument? If we can't get a value of `!`, then how can we call this function? And if we can't call this function, what is it for?

(really sorry, I told you I wasn't the most qualified reviewer :cold_sweat: )

Reply:
> If we can't get a value of !, then how can we call this function? And if we can't call this function, what is it for?

We can't call this function indeed. :)
It's a trait method, and the type of `fn_val` is an associated type. So there might be (and, in fact, there are) other implementations of the trait that pick a non-empty type, and then `call_extra_fn` takes that type. For example [Miri instantiates this trait](https://github.com/rust-lang/miri/pull/808/files#diff-3ee102ac08bbe90033a2dfbf62546499), and there `call_extra_fn` looks like
```rust
    fn call_extra_fn(
        ecx: &mut InterpretCx<'mir, 'tcx, Self>,
        fn_val: Dlsym,
        args: &[OpTy<'tcx, Tag>],
        dest: Option<PlaceTy<'tcx, Tag>>,
        ret: Option<mir::BasicBlock>,
    ) -> InterpResult<'tcx>
```
Notice the `Dlsym` instead of the `!`.

By using `!` in the `const_eval` instance, I basically opt-out of the ability (that the `Machine` trait offers) to have "other" function values, and because there are no "other" function values, `call_extra_fn` does not have to do anything. This surfaces in the type of the method as you said; so we can implement the method trivially without even panicking or so, just by matching on `fn_val` to "prove" to the compiler that we are in unreachable code.

Reply:
> (really sorry, I told you I wasn't the most qualified reviewer cold_sweat )

We could ask @eddyb to also take a look if you want -- but it#s hard to get a time slice from his scheduler. ;)